crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -433,114 +518,47 @@ fn sort_systems(systems: &mut Vec<impl SystemContainer>) -> Result<(), Dependenc
-            .insert(index);
-    }
-    let mut graph = HashMap::with_capacity_and_hasher(systems.len(), Default::default());
-    for (system_index, container) in systems.iter().enumerate() {
-        let dependencies = graph.entry(system_index).or_insert_with(HashMap::default);
-        for label in container.after() {
-            match labelled_systems.get(label) {
-                Some(new_dependencies) => {
-                    for dependency in new_dependencies.ones() {
-                        dependencies
-                            .entry(dependency)
-                            .or_insert_with(HashSet::default)
-                            .insert(label.clone());
-                    }
-                }
-                None => warn!(
-                    "System {} wants to be after unknown system label: {:?}",
-                    systems[system_index].name(),
-                    label
-                ),
-            }
-        }
-        for label in container.before() {
-            match labelled_systems.get(label) {
-                Some(dependants) => {
-                    for dependant in dependants.ones() {
-                        graph
-                            .entry(dependant)
-                            .or_insert_with(HashMap::default)
-                            .entry(system_index)
-                            .or_insert_with(HashSet::default)
-                            .insert(label.clone());
-                    }
-                }
-                None => warn!(
-                    "System {} wants to be before unknown system label: {:?}",
-                    systems[system_index].name(),
+/// Deduplicates and sorts run criteria, and populates resolved input-criteria for piping.
+/// Returns a map of run criteria labels to their indices.
+fn process_run_criteria(
+    run_criteria: &mut Vec<RunCriteriaContainer>,
+) -> Result<
+    HashMap<BoxedRunCriteriaLabel, usize>,
+    DependencyGraphError<HashSet<BoxedRunCriteriaLabel>>,
+> {
+    // This throws away all but the first criteria with identical labels.
+    *run_criteria = run_criteria
+        .drain(..)
+        .map(|container| (container.label.clone(), container))
+        .collect::<HashMap<_, _>>()
+        .drain()
+        .map(|(_, container)| container)
+        .collect();
+    let graph = graph_utils::build_dependency_graph(run_criteria);
+    let order = graph_utils::topological_order(&graph)?;
+    let mut order_inverted = order.iter().enumerate().collect::<Vec<_>>();
+    order_inverted.sort_unstable_by_key(|(_, &key)| key);
+    let labels: HashMap<_, _> = run_criteria
+        .iter()
+        .enumerate()
+        .map(|(index, criteria)| (criteria.label.clone(), order_inverted[index].0))
+        .collect();
+    for criteria in run_criteria.iter_mut() {
+        if let RunCriteriaInner::Chained { input: parent, .. } = &mut criteria.inner {
+            let label = &criteria.after[0];
```

Comment:
This should be commented to explain why only one label is used (assuming we stick with this behavior).

Reply:
Or expand on the comment above, on line 529.

Reply:
This bit is not about that: for piped criteria (I forgot to rename the enums) the first label in `after` is always the criteria to pipe from. [This](https://github.com/Ratysz/bevy/blob/e48811b294e7705496ea0dde038b765ea0d09be6/crates/bevy_ecs/src/schedule/run_criteria.rs#L191) is where it should be documented.